src/smc-webapp/library.cjsx

```diff
@@ -0,0 +1,284 @@
+{Markdown, Space, TimeAgo, ErrorDisplay, Icon, Loading, TimeAgo, Tip, Space} = require('./r_misc')
+{webapp_client} = require('./webapp_client')
+{COLORS}        = require('smc-util/theme')
+
+# src: where the library files are
+# start: open this file after copying the directory
+exports.LIBRARY = LIBRARY =
+    first_steps :
+        src    : '/ext/library/first-steps/src'
+        start  : 'first-steps.tasks'
+
+# used for some styles
+HEIGHT = '275px'
+
+# https://github.com/sagemathinc/cocalc-examples
+exports.examples_path = ROOT = '/ext/library/cocalc-examples'
+
+
+# This is the main library component. It consists of a "selector" and a preview.
+# Later, we probably want to filter by tags, free-text search, ... but for now we just have the main categories
+exports.Library = rclass ({name}) ->
+    displayName : "Library-#{name}"
+
+    reduxProps :
+        "#{name}" :
+            current_path        : rtypes.string
+            library             : rtypes.immutable.Map
+            library_docs_sorted : rtypes.immutable.List
+            library_selected    : rtypes.immutable.Map
+            library_copy_active : rtypes.bool
+        projects:
+            project_map         : rtypes.immutable
+
+    propTypes :
+        actions  : rtypes.object.isRequired
+
+    getInitialState: ->
+        lang        : 'python'
+        show_thumb  : false
+        # sorted_docs : undefined    # depends('library') ->
+
+    metadata: ->
+        @props.library?.getIn(['examples', 'metadata'])
+
+    # this might be better off in actions. the purpose is to prepare the target for the rsync operation.
+    # so far, this works well for all directories -- marked by a "/" at the end.
+    target_path: ->
+        doc = @props.library_selected
+        src = doc.get('src')
+        if doc.get('subdir')
+            subdir = doc.get('subdir')
+        else
+            # directory? cut off the trailing slash
+            if src[src.length - 1] == '/'
+                src = src[...-1]
+                # subdir in current path is the name of the directory
+                subdir = misc.path_split(src).tail
+            else    # otherwise, we're about to copy over a single file → no subdirectory!
+                subdir = ''
+        target = os_path.join(@props.current_path, subdir)
+        #if DEBUG then console.log("copy from", doc.src, "to", target)
+        return target
+
+    # This is the core part of all this: copy over the directory (TODO: a single file)
+    # from the global read-only dir to the user's current directory
+    copy: (doc) ->
+        @props.actions.set_library_copy_active(false)
+        doc = @props.library_selected
+        @props.actions.copy_from_library
+            src    : doc.get('src')
+            target : @target_path()
+            title  : doc.get('title')
+            docid  : doc.get('id')
+            start  : doc?.get('start') ? '/'
+            cb     : => @props.actions.set_library_copy_active(true)
```

Comment:
I read `library_copy_active` as "The library is actively copying" but reading the places where the condition is used, it seems to mean "The copy button is active". I recommend changing the name to `library_is_copying` and swapping the truthiness check in relevant situations. 

It could be `library_copy_allowed` but I think that condition might rely on other data and is more likely to muck up the data semantics.